scheduler/preemption.go

```diff
@@ -435,6 +438,52 @@ OUTER:
 	return filteredBestAllocs
 }
 
+// PreemptForDevice tries to find allocations to preempt to meet devices needed
+// This is called once per task when assigning devices to the task
+func (p *Preemptor) PreemptForDevice(ask *structs.RequestedDevice, devAlloc *deviceAllocator) []*structs.Allocation {
+	// First group allocations by priority
+	allocsByPriority := filterAndGroupPreemptibleAllocs(p.jobPriority, p.currentAllocs)
+	neededCount := ask.Count
+	preemptedCount := 0
+	var preemptedAllocs []*structs.Allocation
+
+	for _, grpAllocs := range allocsByPriority {
+		for _, alloc := range grpAllocs.allocs {
+			for _, tr := range alloc.AllocatedResources.Tasks {
+				if len(tr.Devices) > 0 {
+					// Go through each assigned device group
+					for _, device := range tr.Devices {
+						devID := device.ID()
+
+						// Look up the device instance from the device allocator
+						devInst := devAlloc.Devices[*devID]
```

Comment:
Hmm I think this needs to look a bit more like the network preemptor. Essentially the devInst is like the network device. Preemptions across devices do not help. All devices must be allocated from one device group per request.

The current code could accumulate preemptions across devices in a way that is not productive.

```
resources {
  device "nvidia/gpu" {
    count = 4
  }
}
```

Essentially all of the Nvidia gpus need to be from the same device group. Imagine the node had two types of Nvidia gpus with the following free: 
`nvidia/gpu/1080ti`: 2
`nvidia/gpu/2080ti`: 4

Currently we would add 2 of the 1080ti's and then 2 more of the 2080tis and say we are done really it needs to skip the 1080ti.

Reply:
IS the deviceID the correct way to identify a "device group". IIUC that's the unique ID of the device that we can use to bucket a list of allocations that share a device.